content/en/platform/corda/5.0/developing-applications/api/api-crypto-extensions.md

```diff
@@ -0,0 +1,230 @@
+{{< /note >}}
+
+Double SHA-256 is supported by the platform but let us assume that you want to support Triple SHA-256 where the first pass calculates the message digest and subsequent passes calculate the digest of the previous pass result. The following sections show how to acheive this.
+
+## TripleSha256Digest.kt
+{{< tabs name="TripleSha256Digest">}}
+{{% tab name="Kotlin" %}}
+```kotlin
+package com.example.crypto
+
+import net.corda.v5.crypto.DigestAlgorithmName
+import net.corda.v5.crypto.extensions.DigestAlgorithm
+import java.io.InputStream
+import java.security.MessageDigest
+
+class TripleSha256Digest : DigestAlgorithm {
+    companion object {
+        const val ALGORITHM = "SHA-256-TRIPLE"
+        const val STREAM_BUFFER_SIZE = DEFAULT_BUFFER_SIZE
+    }
+
+    override fun getAlgorithm() = ALGORITHM
+    override fun getDigestLength() = 32
+    override fun digest(bytes: ByteArray): ByteArray = bytes.sha256Bytes().sha256Bytes().sha256Bytes()
+    override fun digest(inputStream: InputStream): ByteArray {
+        val messageDigest = MessageDigest.getInstance(DigestAlgorithmName.SHA2_256.name)
+        val buffer = ByteArray(STREAM_BUFFER_SIZE)
+        while (true) {
+            val read = inputStream.read(buffer)
+            if (read <= 0) break
+            messageDigest.update(buffer, 0, read)
+        }
+        return messageDigest.digest().sha256Bytes().sha256Bytes()
+    }
+}
+
+private fun ByteArray.sha256Bytes(): ByteArray =
+    MessageDigest.getInstance(DigestAlgorithmName.SHA2_256.name).digest(this)
+```
+{{% /tab %}}
+{{% tab name="Java" %}}
+```java
+package com.example.crypto;
+
+import net.corda.v5.crypto.DigestAlgorithmName;
+import net.corda.v5.crypto.extensions.DigestAlgorithm;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class TripleSha256Digest implements DigestAlgorithm {
+    public static String ALGORITHM = "SHA-256-TRIPLE";
+    public static int STREAM_BUFFER_SIZE = 8192;
+
+    @NotNull
+    @Override
+    public String getAlgorithm() {
+        return ALGORITHM;
+    }
+
+    @Override
+    public int getDigestLength() {
+        return 32;
+    }
+
+    @NotNull
+    @Override
+    public byte[] digest(@NotNull byte[] bytes) {
+        try {
+            return sha256Bytes(sha256Bytes(sha256Bytes(bytes)));
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @NotNull
+    @Override
+    public byte[] digest(@NotNull InputStream inputStream) {
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(DigestAlgorithmName.SHA2_256.getName());
+            byte[] buffer = new byte[STREAM_BUFFER_SIZE];
+            while (true) {
+                int read = inputStream.read(buffer);
+                if (read <= 0) break;
```

Comment:
Isn't `0` a legal number of bytes to read? According to the [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-byte:A-):
> Returns:
    the total number of bytes read into the buffer, or -1 if there is no more data because the end of the stream has been reached.

For example, a non-blocking `InputStream` may still be open but not have any more data to return yet.